File: android/src/main/kotlin/com/changteng/agora_rtc/AgoraRtcEngineEventHandler.kt
package com.changteng.agora_rtc

import io.agora.rtc2.IRtcEngineEventHandler

class AgoraRtcEngineEventHandler(
  private val emitter: AgoraEventEmitter
) : IRtcEngineEventHandler() {
  override fun onJoinChannelSuccess(channel: String?, uid: Int, elapsed: Int) {
    emitter.emit(
      "onJoinChannelSuccess",
      mapOf(
        "channel" to (channel ?: ""),
        "uid" to uid,
        "elapsed" to elapsed
      )
    )
  }

  override fun onRejoinChannelSuccess(channel: String?, uid: Int, elapsed: Int) {
    emitter.emit(
      "onRejoinChannelSuccess",
      mapOf(
        "channel" to (channel ?: ""),
        "uid" to uid,
        "elapsed" to elapsed
      )
    )
  }

  override fun onLeaveChannel(stats: IRtcEngineEventHandler.RtcStats?) {
    val payload = mutableMapOf<String, Any?>()
    if (stats != null) {
      payload["duration"] = stats.totalDuration
      payload["txBytes"] = stats.txBytes
      payload["rxBytes"] = stats.rxBytes
      payload["txKBitRate"] = stats.txKBitRate
      payload["rxKBitRate"] = stats.rxKBitRate
      payload["txAudioBytes"] = stats.txAudioBytes
      payload["rxAudioBytes"] = stats.rxAudioBytes
      payload["txVideoBytes"] = stats.txVideoBytes
      payload["rxVideoBytes"] = stats.rxVideoBytes
    }
    emitter.emit("onLeaveChannel", payload)
  }

  override fun onUserJoined(uid: Int, elapsed: Int) {
    emitter.emit(
      "onUserJoined",
      mapOf(
        "uid" to uid,
        "elapsed" to elapsed
      )
    )
  }

  override fun onUserOffline(uid: Int, reason: Int) {
    emitter.emit(
      "onUserOffline",
      mapOf(
        "uid" to uid,
        "reason" to reason
      )
    )
  }

  override fun onFirstRemoteVideoFrame(uid: Int, width: Int, height: Int, elapsed: Int) {
    emitter.emit(
      "onFirstRemoteVideoFrame",
      mapOf(
        "uid" to uid,
        "channelId" to "",
        "width" to width,
        "height" to height
      )
    )
  }

  override fun onClientRoleChangeFailed(reason: Int, currentRole: Int) {
    emitter.emit(
      "onClientRoleChangeFailed",
      mapOf("reason" to reason)
    )
  }

  override fun onConnectionStateChanged(state: Int, reason: Int) {
    emitter.emit(
      "onConnectionStateChanged",
      mapOf(
        "state" to state,
        "reason" to reason
      )
    )
  }

  override fun onRequestToken() {
    emitter.emit("onRequestToken", emptyMap())
  }

  override fun onTokenPrivilegeWillExpire(token: String?) {
    emitter.emit(
      "onTokenPrivilegeWillExpire",
      mapOf("token" to (token ?: ""))
    )
  }

  override fun onError(err: Int) {
    emitter.emit(
      "onError",
      mapOf(
        "error" to err,
        "message" to ""
      )
    )
  }

  override fun onAudioPublishStateChanged(
    channel: String?,
    oldState: Int,
    newState: Int,
    elapseSinceLastState: Int
  ) {
    emitter.emit(
      "onAudioPublishStateChanged",
      mapOf(
        "channel" to (channel ?: ""),
        "oldState" to oldState,
        "newState" to newState,
        "elapseSinceLastState" to elapseSinceLastState
      )
    )
  }

  override fun onUserMuteVideo(uid: Int, muted: Boolean) {
    emitter.emit(
      "onUserMuteVideo",
      mapOf(
        "channelId" to "",
        "uid" to uid,
        "muted" to muted
      )
    )
  }

  override fun onSnapshotTaken(uid: Int, filePath: String?, width: Int, height: Int, errCode: Int) {
    emitter.emit(
      "onSnapshotTaken",
      mapOf(
        "connection" to mapOf<String, Any?>(),
        "uid" to uid,
        "filePath" to (filePath ?: ""),
        "width" to width,
        "height" to height,
        "errCode" to errCode
      )
    )
  }
}

File: ios/Classes/AgoraRtcEngineEventHandler.swift
import AgoraRtcKit

final class AgoraRtcEngineEventHandler: NSObject, AgoraRtcEngineDelegate {
  private let emitter: AgoraEventEmitter

  init(emitter: AgoraEventEmitter) {
    self.emitter = emitter
    super.init()
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinChannel channel: String, withUid uid: UInt, elapsed: Int) {
    emitter.emit(
      type: "onJoinChannelSuccess",
      data: [
        "channel": channel,
        "uid": Int(uid),
        "elapsed": elapsed,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didRejoinChannel channel: String, withUid uid: UInt, elapsed: Int) {
    emitter.emit(
      type: "onRejoinChannelSuccess",
      data: [
        "channel": channel,
        "uid": Int(uid),
        "elapsed": elapsed,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didLeaveChannelWith stats: AgoraChannelStats) {
    emitter.emit(
      type: "onLeaveChannel",
      data: [
        "duration": stats.duration,
        "txBytes": stats.txBytes,
        "rxBytes": stats.rxBytes,
        "txAudioBytes": stats.txAudioBytes,
        "rxAudioBytes": stats.rxAudioBytes,
        "txVideoBytes": stats.txVideoBytes,
        "rxVideoBytes": stats.rxVideoBytes,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didJoinedOfUid uid: UInt, elapsed: Int) {
    emitter.emit(
      type: "onUserJoined",
      data: [
        "uid": Int(uid),
        "elapsed": elapsed,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didOfflineOfUid uid: UInt, reason: AgoraUserOfflineReason) {
    emitter.emit(
      type: "onUserOffline",
      data: [
        "uid": Int(uid),
        "reason": reason.rawValue,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, firstRemoteVideoFrameOfUid uid: UInt, size: CGSize, elapsed: Int) {
    emitter.emit(
      type: "onFirstRemoteVideoFrame",
      data: [
        "uid": Int(uid),
        "channelId": "",
        "width": Int(size.width),
        "height": Int(size.height),
      ]
    )
  }

  func rtcEngine(
    _ engine: AgoraRtcEngineKit,
    didClientRoleChangeFailed reason: AgoraClientRoleChangeFailedReason,
    currentRole: AgoraClientRole
  ) {
    emitter.emit(
      type: "onClientRoleChangeFailed",
      data: [
        "reason": reason.rawValue,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, connectionChangedTo state: AgoraConnectionState, reason: AgoraConnectionChangedReason) {
    emitter.emit(
      type: "onConnectionStateChanged",
      data: [
        "state": state.rawValue,
        "reason": reason.rawValue,
      ]
    )
  }

  func rtcEngineRequestToken(_ engine: AgoraRtcEngineKit) {
    emitter.emit(type: "onRequestToken", data: [:])
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, tokenPrivilegeWillExpire token: String) {
    emitter.emit(
      type: "onTokenPrivilegeWillExpire",
      data: ["token": token]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didOccurError errorCode: AgoraErrorCode) {
    emitter.emit(
      type: "onError",
      data: [
        "error": errorCode.rawValue,
        "message": "",
      ]
    )
  }

  func rtcEngine(
    _ engine: AgoraRtcEngineKit,
    didAudioPublishStateChange channel: String,
    oldState: AgoraStreamPublishState,
    newState: AgoraStreamPublishState,
    elapseSinceLastState: Int
  ) {
    emitter.emit(
      type: "onAudioPublishStateChanged",
      data: [
        "channel": channel,
        "oldState": oldState.rawValue,
        "newState": newState.rawValue,
        "elapseSinceLastState": elapseSinceLastState,
      ]
    )
  }

  func rtcEngine(_ engine: AgoraRtcEngineKit, didVideoMuted muted: Bool, byUid uid: UInt) {
    emitter.emit(
      type: "onUserMuteVideo",
      data: [
        "channelId": "",
        "uid": Int(uid),
        "muted": muted,
      ]
    )
  }

  func rtcEngine(
    _ engine: AgoraRtcEngineKit,
    snapshotTaken uid: UInt,
    filePath: String,
    width: Int,
    height: Int,
    errCode: Int
  ) {
    emitter.emit(
      type: "onSnapshotTaken",
      data: [
        "connection": [:],
        "uid": Int(uid),
        "filePath": filePath,
        "width": width,
        "height": height,
        "errCode": errCode,
      ]
    )
  }
}
